src/soc/hw/noc/test/test_noc_buffer.py

```diff
@@ -0,0 +1,343 @@
+def write_buffer(dut, clk, width, max_delay=5, write_items=1000, write_random_data=True):
+    """
+    Before writing wait between 0 and max_delay cycles.
+    """
+
+    global _buffer_flits, _buffer_last
+
+    wrcnt = 0
+    while True:
+        # Insert random wait before the next write
+        wr_delay = random.randint(0, max_delay)
+        dut._log.debug("WRITE: Wait for {} clock cycles".format(wr_delay))
+        for _ in range(wr_delay):
+            dut.in_valid <= 0
+            yield RisingEdge(clk)
+
+        # Generate and write flit, keep track of it for checking
+        if write_random_data:
+            flit = random.getrandbits(width)
+            last = random.getrandbits(1)
+        else:
+            flit = wrcnt % (2 ** width)
+            last = wrcnt % 2
+        dut.in_flit <= flit
+        dut.in_last <= last
+        dut.in_valid <= 1
+
+        yield RisingEdge(clk)
+
+        # Wait until buffer is ready
+        if dut.in_ready != 1:
+            while dut.in_ready.value.integer != 1:
+                yield RisingEdge(clk)
+
+        _buffer_flits.append(flit)
+        _buffer_last.append(last)
+        wrcnt += 1
+        dut._log.debug("WRITE: Wrote word {} to buffer, value {}, last {}".format(wrcnt, hex(flit), last))
+
+        if wrcnt >= write_items:
+            dut.in_valid <= 0
+            return
+
+
+@cocotb.coroutine
+def read_buffer(dut, clk, max_delay=5, read_items=1000):
+    """
+    Before the read starts a delay between 0 and max_delay cycles is
+    waited.
+    """
+
+    global _buffer_flits, _buffer_last
+
+    rdcnt = 0
+    while True:
+        # Determine wait cycles before next read
+        rd_delay = random.randint(0, max_delay)
+        dut._log.debug("READ: Wait for {} clock cycles".format(rd_delay))
+        for _ in range(rd_delay):
+            yield RisingEdge(clk)
+        else:
+            dut.out_ready <= 1
+            yield RisingEdge(clk)
+
+        while dut.out_valid.value.integer:
+            # Get current flit and flag
+            flit = dut.out_flit.value.integer
+            last = dut.out_last.value.integer
+            flit_expected = _buffer_flits.pop(0)
+            last_expected = _buffer_last.pop(0)
+            rdcnt += 1
+            dut._log.debug("READ: Read flit {} in read {}.".format(hex(flit), rdcnt))
+
+            # Check read flit and flag
+            if flit_expected != flit or last_expected != last:
+                raise TestFailure("READ: Flit expected {}, read {}, ".format(hex(flit_expected), hex(flit)) +\
+                                  "last expected {}, read {}, at read {}.".format(last_expected, last, rdcnt))
+
+            yield RisingEdge(clk)
+
+        else:
+            dut._log.debug("READ: Buffer empty, not reading")
+            dut.out_ready <= 0
+
+        if rdcnt >= read_items:
+            return
+
+
+@cocotb.coroutine
+def fullpacket_test_case(dut):
+    """
+    Special test case to test FULLPACKET behavior.
```

Comment:
nit: s/Special/Directed

Maybe give the required width/depth parameters here as well?